File: native/ios-app/Sources/FireAppSession/FireWebViewLoginCoordinator.swift
import Foundation
import WebKit

enum FireBootstrapHTMLHeuristics {
    static let reusableLoginBootstrapScoreThreshold = 8

    static func preferredHTML(
        browserFetchedHomeHTML: String?,
        currentPageHTML: String?
    ) -> String? {
        let homeScore = score(browserFetchedHomeHTML)
        let currentScore = score(currentPageHTML)

        if homeScore > currentScore, let browserFetchedHomeHTML = nonEmpty(browserFetchedHomeHTML) {
            return browserFetchedHomeHTML
        }

        if let currentPageHTML = nonEmpty(currentPageHTML) {
            return currentPageHTML
        }

        return nonEmpty(browserFetchedHomeHTML)
    }

    static func score(_ html: String?) -> Int {
        guard let html else {
            return 0
        }

        let normalized = html.lowercased()
        var score = 0

        if normalized.contains("id=\"data-discourse-setup\"")
            || normalized.contains("id='data-discourse-setup'")
            || normalized.contains("data-preloaded")
        {
            score += 8
        }
        if normalized.contains("meta name=\"shared_session_key\"")
            || normalized.contains("meta name='shared_session_key'")
        {
            score += 4
        }
        if normalized.contains("\"long_polling_base_url\"") {
            score += 6
        }
        if normalized.contains("\"topictrackingstatemeta\"") {
            score += 5
        }
        if normalized.contains("\"notification_channel_position\"") {
            score += 4
        }
        if normalized.contains("\"categories\":") {
            score += 3
        }
        if normalized.contains("\"top_tags\":") {
            score += 3
        }
        if normalized.contains("\"can_tag_topics\"") {
            score += 2
        }
        if normalized.contains("\"sitesettings\":") {
            score += 2
        }
        if normalized.contains("meta name=\"current-username\"")
            || normalized.contains("meta name='current-username'")
        {
            score += 2
        }
        if normalized.contains("meta name=\"csrf-token\"")
            || normalized.contains("meta name='csrf-token'")
        {
            score += 1
        }

        return score
    }

    static func isReusableLoginBootstrap(_ html: String?) -> Bool {
        score(html) >= reusableLoginBootstrapScoreThreshold
    }

    private static func nonEmpty(_ html: String?) -> String? {
        guard let html, !html.isEmpty else {
            return nil
        }
        return html
    }
}

enum FireBootstrapHTMLMetadataParser {
    static func currentUsername(from html: String?) -> String? {
        metaContent(named: "current-username", in: html)
    }

    static func csrfToken(from html: String?) -> String? {
        metaContent(named: "csrf-token", in: html)
    }

    private static func metaContent(named name: String, in html: String?) -> String? {
        guard let html, !html.isEmpty else {
            return nil
        }

        let escapedName = NSRegularExpression.escapedPattern(for: name)
        let patterns = [
            #"<meta\b[^>]*\bname\s*=\s*["']\#(escapedName)["'][^>]*\bcontent\s*=\s*["']([^"']+)["'][^>]*>"#,
            #"<meta\b[^>]*\bcontent\s*=\s*["']([^"']+)["'][^>]*\bname\s*=\s*["']\#(escapedName)["'][^>]*>"#,
        ]

        for pattern in patterns {
            guard
                let regex = try? NSRegularExpression(
                    pattern: pattern,
                    options: [.caseInsensitive]
                )
            else {
                continue
            }

            let range = NSRange(html.startIndex..<html.endIndex, in: html)
            guard
                let match = regex.firstMatch(in: html, options: [], range: range),
                match.numberOfRanges > 1,
                let contentRange = Range(match.range(at: 1), in: html)
            else {
                continue
            }

            let content = html[contentRange].trimmingCharacters(in: .whitespacesAndNewlines)
            if !content.isEmpty {
                return content
            }
        }

        return nil
    }
}

public struct FireLoginSyncReadiness: Sendable, Equatable {
    public let isReady: Bool
    public let username: String?
    public let hasAuthCookies: Bool
    public let hasBootstrapHTML: Bool
    public let preferredBootstrapScore: Int

    public init(
        isReady: Bool,
        username: String?,
        hasAuthCookies: Bool,
        hasBootstrapHTML: Bool,
        preferredBootstrapScore: Int
    ) {
        self.isReady = isReady
        self.username = username
        self.hasAuthCookies = hasAuthCookies
        self.hasBootstrapHTML = hasBootstrapHTML
        self.preferredBootstrapScore = preferredBootstrapScore
    }
}

public enum FireWebViewLoginCoordinatorError: LocalizedError {
    case loginSyncNotReady(FireLoginSyncReadiness)

    public var errorDescription: String? {
        switch self {
        case let .loginSyncNotReady(readiness):
            if !readiness.hasAuthCookies {
                return "登录状态尚未写入站点 Cookie，请稍后再试。"
            }
            if !readiness.hasBootstrapHTML {
                return "登录页还未拿到可用的站点引导数据，请稍后再试。"
            }
            return "登录状态尚未准备完成，请稍后再试。"
        }
    }
}

protocol FireLoginSessionStoring: Sendable {
    func currentSessionSnapshot() async throws -> SessionState
    func restorePersistedSessionIfAvailable() async throws -> SessionState?
    func finalizeLoginFromWebView(
        _ captured: FireCapturedLoginState,
        allowLowConfidenceSessionCookies: Bool
    ) async throws -> LoginFinalizationResultState
    func syncLoginContext(_ captured: FireCapturedLoginState) async throws -> SessionState
    func refreshBootstrapIfNeeded() async throws -> SessionState
    func refreshCsrfTokenIfNeeded() async throws -> SessionState
    func logout() async throws -> SessionState
    func logoutLocal(preserveCfClearance: Bool) async throws -> SessionState
    func applyPlatformCookies(_ cookies: [PlatformCookieState]) async throws -> SessionState
}

extension FireSessionStore: FireLoginSessionStoring {
    func currentSessionSnapshot() async throws -> SessionState {
        try snapshot()
    }
}

@MainActor
public final class FireWebViewLoginCoordinator {
    private let sessionStore: any FireLoginSessionStoring

    public init(sessionStore: FireSessionStore) {
        self.sessionStore = sessionStore
    }

    init(
        loginSessionStore: any FireLoginSessionStoring
    ) {
        self.sessionStore = loginSessionStore
    }

    public func restorePersistedSessionIfAvailable() async throws -> SessionState? {
        try await sessionStore.restorePersistedSessionIfAvailable()
    }

    public func completeLogin(
        from webView: WKWebView
    ) async throws -> SessionState {
        let captured = try await captureLoginState(from: webView)
        let readiness = loginSyncReadiness(for: captured)
        guard readiness.isReady else {
            throw FireWebViewLoginCoordinatorError.loginSyncNotReady(readiness)
        }
        return try await completeLogin(captured)
    }

    func completeLogin(
        _ captured: FireCapturedLoginState
    ) async throws -> SessionState {
        let finalized = try await sessionStore.finalizeLoginFromWebView(
            captured,
            allowLowConfidenceSessionCookies: true
        )
        if finalized.session.loginPhase == .ready {
            return finalized.session
        }

        return try await sessionStore.refreshBootstrapIfNeeded()
    }

    public func logout() async throws -> SessionState {
        let state: SessionState
        do {
            state = try await sessionStore.logout()
        } catch {
            state = try await sessionStore.logoutLocal(preserveCfClearance: true)
        }
        try await clearSameSitePlatformCookies(preserving: ["cf_clearance"])
        return state
    }

    public func captureLoginState(from webView: WKWebView) async throws -> FireCapturedLoginState {
        let currentURL = webView.url?.absoluteString
        async let username = readCurrentUsername(in: webView)
        async let csrfToken = readCsrfToken(in: webView)
        async let preloadedHTML = readStringJavaScript(
            script: FireLoginScripts.readPreloadedData,
            in: webView
        )
        async let browserUserAgent = readStringJavaScript(
            script: "navigator.userAgent",
            in: webView
        )
        async let cookies = relevantCookies(from: webView)

        let capturedUsername = try await username
        let capturedCsrfToken = try await csrfToken
        let capturedPreloadedHTML = try await preloadedHTML
        let capturedBrowserUserAgent = try await browserUserAgent
        let capturedCookies = try await cookies
        let resolvedUsername =
            capturedUsername
            ?? FireBootstrapHTMLMetadataParser.currentUsername(from: capturedPreloadedHTML)
        let resolvedCsrfToken =
            capturedCsrfToken
            ?? FireBootstrapHTMLMetadataParser.csrfToken(from: capturedPreloadedHTML)

        return FireCapturedLoginState(
            currentURL: currentURL,
            username: resolvedUsername,
            csrfToken: resolvedCsrfToken,
            homeHTML: capturedPreloadedHTML,
            browserUserAgent: capturedBrowserUserAgent,
            cookies: capturedCookies
        )
    }

    public func platformCookiesForSessionResync() async throws -> [PlatformCookieState] {
        try await relevantCookies(
            from: WKWebsiteDataStore.default().httpCookieStore
        )
    }

    public func refreshPlatformCookies() async throws -> SessionState {
        let cookies = try await platformCookiesForSessionResync()
        return try await applyPlatformCookiesIfAuthoritative(cookies)
    }

    func applyPlatformCookiesIfAuthoritative(
        _ cookies: [PlatformCookieState]
    ) async throws -> SessionState {
        guard Self.containsActiveAuthCookies(in: cookies) else {
            return try await sessionStore.currentSessionSnapshot()
        }
        return try await sessionStore.applyPlatformCookies(cookies)
    }

    public func probeLoginSyncReadiness(from webView: WKWebView) async throws -> FireLoginSyncReadiness {
        async let username = readCurrentUsername(in: webView)
        async let preloadedHTML = readStringJavaScript(
            script: FireLoginScripts.readPreloadedData,
            in: webView
        )
        async let cookies = relevantCookies(from: webView)

        let capturedUsername = try await username
        let capturedPreloadedHTML = try await preloadedHTML
        let capturedCookies = try await cookies
        return loginSyncReadiness(
            username: capturedUsername
                ?? FireBootstrapHTMLMetadataParser.currentUsername(from: capturedPreloadedHTML),
            cookies: capturedCookies,
            preferredBootstrapScore: FireBootstrapHTMLHeuristics.score(capturedPreloadedHTML)
        )
    }

    public func logoutLocalAndClearPlatformCookies(
        preserveCfClearance: Bool = true
    ) async throws -> SessionState {
        let state = try await sessionStore.logoutLocal(preserveCfClearance: preserveCfClearance)
        try await clearSameSitePlatformCookies(
            preserving: preserveCfClearance ? ["cf_clearance"] : []
        )
        return state
    }

    private func relevantCookies(from webView: WKWebView) async throws -> [PlatformCookieState] {
        try await relevantCookies(from: webView.configuration.websiteDataStore.httpCookieStore)
    }

    private func relevantCookies(from store: WKHTTPCookieStore) async throws -> [PlatformCookieState] {
        let allCookies = try await httpCookies(from: store)
        var relevant: [PlatformCookieState] = []
        for cookie in allCookies {
            guard cookie.domain.range(of: "linux.do", options: .caseInsensitive) != nil else {
                continue
            }
            guard isActiveCookie(cookie) else {
                continue
            }

            relevant.append(
                PlatformCookieState(
                    name: cookie.name,
                    value: cookie.value,
                    domain: cookie.domain,
                    path: cookie.path,
                    expiresAtUnixMs: cookie.expiresDate.map { Int64($0.timeIntervalSince1970 * 1000) },
                    sameSite: nil
                )
            )
        }
        return relevant
    }

    private func isActiveCookie(_ cookie: HTTPCookie) -> Bool {
        let normalizedValue = cookie.value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedValue.isEmpty else {
            return false
        }
        guard let expiresDate = cookie.expiresDate else {
            return true
        }
        return expiresDate > Date()
    }

    private func httpCookies(from store: WKHTTPCookieStore) async throws -> [HTTPCookie] {
        try await withCheckedThrowingContinuation { continuation in
            store.getAllCookies { cookies in
                continuation.resume(returning: cookies)
            }
        }
    }

    private func clearSameSitePlatformCookies(
        preserving preservedCookieNames: Set<String> = []
    ) async throws {
        try await clearPlatformCookies { cookie in
            isSameSiteCookie(cookie) && !preservedCookieNames.contains(cookie.name)
        }
    }

    private func clearPlatformCookies(
        matching shouldDelete: (HTTPCookie) -> Bool
    ) async throws {
        let webKitStore = WKWebsiteDataStore.default().httpCookieStore
        let webKitCookies = try await httpCookies(from: webKitStore)
        for cookie in webKitCookies where shouldDelete(cookie) {
            await deleteCookie(cookie, from: webKitStore)
        }

        for cookie in HTTPCookieStorage.shared.cookies ?? [] where shouldDelete(cookie) {
            HTTPCookieStorage.shared.deleteCookie(cookie)
        }
    }

    private func isSameSiteCookie(_ cookie: HTTPCookie) -> Bool {
        cookie.domain.range(of: "linux.do", options: .caseInsensitive) != nil
    }

    private func deleteCookie(_ cookie: HTTPCookie, from store: WKHTTPCookieStore) async {
        await withCheckedContinuation { continuation in
            store.delete(cookie) {
                continuation.resume()
            }
        }
    }

    private func readCurrentUsername(in webView: WKWebView) async throws -> String? {
        try await readStringJavaScript(
            script: """
            (function() {
              try {
                var meta = document.querySelector('meta[name="current-username"]');
                if (meta && meta.content) return meta.content;
                if (
                  typeof Discourse !== 'undefined'
                  && Discourse.User
                  && typeof Discourse.User.current === 'function'
                ) {
                  var currentUser = Discourse.User.current();
                  if (currentUser && currentUser.username) return currentUser.username;
                }
              } catch (error) {}
              return null;
            })();
            """,
            in: webView
        )
    }

    private func readCsrfToken(in webView: WKWebView) async throws -> String? {
        try await readStringJavaScript(
            script: """
            (function() {
              var meta = document.querySelector('meta[name="csrf-token"]');
              if (meta && meta.content) return meta.content;
              return null;
            })();
            """,
            in: webView
        )
    }

    private func readCurrentPageBootstrapScore(in webView: WKWebView) async throws -> Int {
        try await readIntJavaScript(
            script: """
            (function() {
              try {
                var score = 0;
                if (document.querySelector('#data-discourse-setup,[data-preloaded]')) score += 8;
                if (document.querySelector('meta[name="shared_session_key"]')) score += 4;
                if (document.querySelector('meta[name="current-username"]')) score += 2;
                if (document.querySelector('meta[name="csrf-token"]')) score += 1;
                return score;
              } catch (error) {
                return 0;
              }
            })();
            """,
            in: webView
        ) ?? 0
    }

    private func readStringJavaScript(script: String, in webView: WKWebView) async throws -> String? {
        let value = try await evaluateJavaScript(script: script, in: webView)

        guard let string = value as? String, !string.isEmpty, string != "null" else {
            return nil
        }
        return string
    }

    private func readIntJavaScript(script: String, in webView: WKWebView) async throws -> Int? {
        let value = try await evaluateJavaScript(script: script, in: webView)

        if let number = value as? NSNumber {
            return number.intValue
        }
        if let string = value as? String {
            return Int(string)
        }
        return nil
    }

    private func evaluateJavaScript(script: String, in webView: WKWebView) async throws -> Any? {
        try await withCheckedThrowingContinuation {
            (continuation: CheckedContinuation<Any?, Error>) in
            webView.evaluateJavaScript(script) { result, error in
                if let error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume(returning: result)
                }
            }
        }
    }

    func loginSyncReadiness(for captured: FireCapturedLoginState) -> FireLoginSyncReadiness {
        return loginSyncReadiness(
            username: captured.username,
            cookies: captured.cookies,
            preferredBootstrapScore: FireBootstrapHTMLHeuristics.score(captured.homeHTML)
        )
    }

    func loginSyncReadiness(
        username: String?,
        cookies: [PlatformCookieState],
        preferredBootstrapScore: Int
    ) -> FireLoginSyncReadiness {
        let normalizedUsername = normalizeUsername(username)
        let hasUsername = !(normalizedUsername?.isEmpty ?? true)
        let hasAuthCookies = containsAuthCookies(in: cookies)
        let hasBootstrapHTML =
            preferredBootstrapScore >= FireBootstrapHTMLHeuristics.reusableLoginBootstrapScoreThreshold

        return FireLoginSyncReadiness(
            isReady: hasUsername && hasAuthCookies && hasBootstrapHTML,
            username: normalizedUsername,
            hasAuthCookies: hasAuthCookies,
            hasBootstrapHTML: hasBootstrapHTML,
            preferredBootstrapScore: preferredBootstrapScore
        )
    }

    private func normalizeUsername(_ username: String?) -> String? {
        let trimmed = username?.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed?.isEmpty == true ? nil : trimmed
    }

    nonisolated static func containsActiveAuthCookies(
        in cookies: [PlatformCookieState]
    ) -> Bool {
        let activeCookies = cookies.filter { cookie in
            let value = cookie.value.trimmingCharacters(in: .whitespacesAndNewlines)
            return !value.isEmpty && !(cookie.expiresAtUnixMs.map { $0 <= currentUnixMs() } ?? false)
        }

        return activeCookies.contains(where: { $0.name == "_t" })
            && activeCookies.contains(where: { $0.name == "_forum_session" })
    }

    private func containsAuthCookies(in cookies: [PlatformCookieState]) -> Bool {
        Self.containsActiveAuthCookies(in: cookies)
    }

    private func isLinuxDoHost(_ url: URL?) -> Bool {
        guard let host = url?.host?.lowercased() else {
            return false
        }

        return host == "linux.do" || host.hasSuffix(".linux.do")
    }

    private nonisolated static func currentUnixMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}

File: native/android-app/src/main/java/com/fire/app/session/FireWebViewLoginCoordinator.kt
package com.fire.app.session

import android.webkit.CookieManager
import android.webkit.WebView
import java.net.URI
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import org.json.JSONObject
import kotlin.coroutines.resume
import uniffi.fire_uniffi_session.CookieSweepPlanState
import uniffi.fire_uniffi_session.LoginPhaseState
import uniffi.fire_uniffi_session.PlatformCookieState
import uniffi.fire_uniffi_session.SessionState
import uniffi.fire_uniffi_session.WebViewCookieActionState
import uniffi.fire_uniffi_session.WebViewCookieInfoState

data class FireLoginSyncReadiness(
    val isReady: Boolean,
    val username: String?,
    val hasAuthCookies: Boolean,
    val hasBootstrapHtml: Boolean,
    val preferredBootstrapScore: Int,
)

class FireWebViewLoginCoordinator(
    private val sessionStore: FireSessionStore,
    private val loginBaseUrl: String = "https://linux.do",
) {
    suspend fun restorePersistedSessionIfAvailable(): SessionState? {
        return sessionStore.restorePersistedSessionIfAvailable()
    }

    suspend fun completeLogin(webView: WebView): SessionState {
        val captured = captureLoginState(webView)
        val readiness = loginSyncReadiness(captured)
        check(readiness.isReady) { "登录状态尚未准备完成" }
        return completeLogin(captured)
    }

    suspend fun completeLogin(captured: FireCapturedLoginState): SessionState {
        val finalization = sessionStore.finalizeLoginFromWebView(
            captured = captured,
            allowLowConfidenceSessionCookies = true,
        )
        if (finalization.session.loginPhase == LoginPhaseState.READY) {
            return finalization.session
        }

        return sessionStore.refreshBootstrapIfNeeded()
    }

    suspend fun logout(): SessionState {
        return sessionStore.logout()
    }

    suspend fun syncBrowserContext(webView: WebView): SessionState {
        return applyPlatformCookiesIfAuthoritative(relevantCookies(webView))
    }

    suspend fun primeCookies(
        webView: WebView,
        targetUrl: String = webView.url?.takeIf { it.isNotBlank() } ?: "$loginBaseUrl/",
    ) {
        val actions = sessionStore.webViewPrimingPayload(targetUrl)
        executeCookieActions(actions)
    }

    suspend fun sweepCookies(
        webView: WebView,
        names: List<String> = SESSION_COOKIE_NAMES,
        targetUrl: String = webView.url?.takeIf { it.isNotBlank() } ?: "$loginBaseUrl/",
    ): List<CookieSweepPlanState> {
        val cookieInfos = webViewCookieInfos(targetUrl)
        return names.map { name ->
            sessionStore.cookieSweepPlan(
                targetUrl = targetUrl,
                name = name,
                webViewCookies = cookieInfos,
            ).also { plan ->
                executeCookieActions(plan.actions)
            }
        }
    }

    suspend fun nuclearResetCookies(
        webView: WebView,
        targetUrl: String = webView.url?.takeIf { it.isNotBlank() } ?: "$loginBaseUrl/",
    ) {
        val plan = sessionStore.cookieNuclearResetPlan(
            targetUrl = targetUrl,
            webViewCookies = webViewCookieInfos(targetUrl),
        )
        executeCookieActions(plan.actions)
    }

    suspend fun webViewCookieInfos(targetUrl: String? = "$loginBaseUrl/"): List<WebViewCookieInfoState> =
        withContext(Dispatchers.Main) {
            relevantCookieUrls(targetUrl).flatMap { url ->
                FireWebViewCookieActionSupport.parseCookieHeader(
                    CookieManager.getInstance().getCookie(url),
                )
            }.distinctBy { "${it.name}\u0000${it.value}\u0000${it.domain}\u0000${it.path}" }
        }

    suspend fun executeCookieActions(actions: List<WebViewCookieActionState>) {
        if (actions.isEmpty()) {
            return
        }
        withContext(Dispatchers.Main) {
            val cookieManager = CookieManager.getInstance()
            for (action in actions) {
                when (action) {
                    is WebViewCookieActionState.SetRaw -> {
                        cookieManager.setCookieSuspend(action.url, action.setCookie)
                    }
                    is WebViewCookieActionState.DeleteExact -> {
                        cookieManager.setCookieSuspend(
                            action.url,
                            FireWebViewCookieActionSupport.expiredCookieHeader(
                                name = action.name,
                                domain = action.domain,
                                path = action.path,
                            ),
                        )
                    }
                    is WebViewCookieActionState.DeleteByName -> {
                        FireWebViewCookieActionSupport.deleteByNameHeaders(
                            url = action.url,
                            name = action.name,
                        ).forEach { header ->
                            cookieManager.setCookieSuspend(action.url, header)
                        }
                    }
                }
            }
            cookieManager.flush()
        }
    }

    suspend fun probeLoginSyncReadiness(webView: WebView): FireLoginSyncReadiness {
        val captured = captureLoginState(webView)
        return loginSyncReadiness(captured)
    }

    suspend fun captureLoginState(webView: WebView): FireCapturedLoginState = withContext(Dispatchers.Main) {
        val currentUrl = webView.url ?: loginBaseUrl
        val usernameJson = webView.evaluateJavascriptSuspend(FireLoginScripts.readCurrentUsername)
        val csrfJson = webView.evaluateJavascriptSuspend(FireLoginScripts.readCsrfToken)
        val preloadedJson = webView.evaluateJavascriptSuspend(FireLoginScripts.readPreloadedData)

        val preloadedHtml = preloadedJson.decodeJsonStringOrNull()
        val resolvedUsername = usernameJson.decodeJsonStringOrNull()
            ?: FireBootstrapHtmlMetadataParser.currentUsername(preloadedHtml)
        val resolvedCsrfToken = csrfJson.decodeJsonStringOrNull()
            ?: FireBootstrapHtmlMetadataParser.csrfToken(preloadedHtml)

        FireCapturedLoginState(
            currentUrl = currentUrl,
            username = resolvedUsername,
            csrfToken = resolvedCsrfToken,
            homeHtml = preloadedHtml,
            browserUserAgent = webView.settings.userAgentString?.takeIf { it.isNotBlank() },
            cookies = relevantCookies(currentUrl),
        )
    }

    private suspend fun relevantCookies(webView: WebView): List<PlatformCookieState> = withContext(Dispatchers.Main) {
        relevantCookies(webView.url)
    }

    private fun relevantCookies(currentUrl: String?): List<PlatformCookieState> {
        val merged = LinkedHashMap<String, PlatformCookieState>()
        for (url in relevantCookieUrls(currentUrl)) {
            CookieManager.getInstance()
                .getCookie(url)
                .orEmpty()
                .let(FireWebViewCookieActionSupport::parsePlatformCookies)
                .forEach { cookie ->
                    if (cookie.name !in merged) {
                        merged[cookie.name] = cookie
                    }
                }
        }
        return merged.values.toList()
    }

    private fun relevantCookieUrls(currentUrl: String?): LinkedHashSet<String> {
        return linkedSetOf<String>().also { urls ->
            currentUrl?.takeIf { it.isNotBlank() }?.let(urls::add)
            urls.add(loginBaseUrl)
            urls.add("$loginBaseUrl/")
        }
    }

    private suspend fun applyPlatformCookiesIfAuthoritative(
        cookies: List<PlatformCookieState>,
    ): SessionState {
        if (!containsActiveAuthCookies(cookies)) {
            return sessionStore.snapshot()
        }
        return sessionStore.applyPlatformCookies(cookies)
    }

    private fun loginSyncReadiness(captured: FireCapturedLoginState): FireLoginSyncReadiness {
        return loginSyncReadiness(
            username = captured.username,
            cookies = captured.cookies,
            preferredBootstrapScore = FireBootstrapHtmlHeuristics.score(captured.homeHtml),
        )
    }

    private fun loginSyncReadiness(
        username: String?,
        cookies: List<PlatformCookieState>,
        preferredBootstrapScore: Int,
    ): FireLoginSyncReadiness {
        val normalizedUsername = username?.trim()?.takeIf { it.isNotEmpty() }
        val hasAuthCookies = containsActiveAuthCookies(cookies)
        val hasBootstrapHtml =
            preferredBootstrapScore >= FireBootstrapHtmlHeuristics.REUSABLE_LOGIN_BOOTSTRAP_SCORE_THRESHOLD
        return FireLoginSyncReadiness(
            isReady = normalizedUsername != null && hasAuthCookies && hasBootstrapHtml,
            username = normalizedUsername,
            hasAuthCookies = hasAuthCookies,
            hasBootstrapHtml = hasBootstrapHtml,
            preferredBootstrapScore = preferredBootstrapScore,
        )
    }

    private suspend fun WebView.evaluateJavascriptSuspend(script: String): String =
        suspendCancellableCoroutine { continuation ->
            evaluateJavascript(script) { value ->
                continuation.resume(value ?: "null")
            }
        }

    private fun String.decodeJsonStringOrNull(): String? {
        if (this == "null") {
            return null
        }

        return runCatching {
            JSONObject("""{"value":$this}""")
                .optString("value")
                .trim()
                .takeIf { it.isNotEmpty() }
        }.getOrNull()
    }

    companion object {
        private val SESSION_COOKIE_NAMES = listOf("_t", "_forum_session", "cf_clearance")

        fun containsActiveAuthCookies(cookies: List<PlatformCookieState>): Boolean {
            val nowUnixMs = System.currentTimeMillis()
            val activeCookies = cookies.filter { cookie ->
                val value = cookie.value.trim()
                value.isNotEmpty() && (cookie.expiresAtUnixMs?.let { it > nowUnixMs } ?: true)
            }
            return activeCookies.any { it.name == "_t" } &&
                activeCookies.any { it.name == "_forum_session" }
        }
    }
}

internal object FireWebViewCookieActionSupport {
    fun parseCookieHeader(header: String?): List<WebViewCookieInfoState> {
        return cookiePairs(header).map { (name, value) ->
            WebViewCookieInfoState(
                name = name,
                value = value,
                domain = null,
                path = null,
                hostOnly = null,
                secure = null,
                httpOnly = null,
                sameSite = null,
                expiresAtUnixMs = null,
            )
        }
    }

    fun parsePlatformCookies(header: String?): List<PlatformCookieState> {
        return cookiePairs(header).map { (name, value) ->
            PlatformCookieState(
                name = name,
                value = value,
                domain = null,
                path = null,
                expiresAtUnixMs = null,
                sameSite = null,
            )
        }
    }

    fun expiredCookieHeader(name: String, domain: String?, path: String): String {
        return buildString {
            append(name)
            append("=; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT")
            append("; Path=")
            append(path.ifBlank { "/" })
            if (!domain.isNullOrBlank()) {
                append("; Domain=")
                append(domain.trim())
            }
        }
    }

    fun deleteByNameHeaders(url: String, name: String): List<String> {
        val domains = linkedSetOf<String?>()
        domains.add(null)
        runCatching { URI(url).host }
            .getOrNull()
            ?.trim()
            ?.lowercase()
            ?.takeIf { it.isNotEmpty() }
            ?.let { host ->
                domains.add(host)
                domains.add(".$host")
                if (host == "linux.do" || host.endsWith(".linux.do")) {
                    domains.add(".linux.do")
                }
            }
        return domains.map { domain ->
            expiredCookieHeader(name = name, domain = domain, path = "/")
        }
    }

    private fun cookiePairs(header: String?): List<Pair<String, String>> {
        return header.orEmpty()
            .split(";")
            .mapNotNull { segment ->
                val trimmed = segment.trim()
                if (trimmed.isEmpty()) {
                    return@mapNotNull null
                }
                val separator = trimmed.indexOf('=')
                if (separator <= 0) {
                    return@mapNotNull null
                }
                val name = trimmed.substring(0, separator).trim()
                val value = trimmed.substring(separator + 1).trim()
                if (name.isEmpty() || value.isEmpty()) {
                    return@mapNotNull null
                }
                name to value
            }
    }
}

private suspend fun CookieManager.setCookieSuspend(url: String, value: String): Boolean =
    suspendCancellableCoroutine { continuation ->
        setCookie(url, value) { accepted ->
            continuation.resume(accepted)
        }
    }

private object FireBootstrapHtmlHeuristics {
    const val REUSABLE_LOGIN_BOOTSTRAP_SCORE_THRESHOLD = 8

    fun score(html: String?): Int {
        if (html.isNullOrBlank()) {
            return 0
        }

        val normalized = html.lowercase()
        var score = 0
        if (
            normalized.contains("id=\"data-discourse-setup\"") ||
            normalized.contains("id='data-discourse-setup'") ||
            normalized.contains("data-preloaded")
        ) {
            score += 8
        }
        if (
            normalized.contains("meta name=\"shared_session_key\"") ||
            normalized.contains("meta name='shared_session_key'")
        ) {
            score += 4
        }
        if (
            normalized.contains("meta name=\"current-username\"") ||
            normalized.contains("meta name='current-username'")
        ) {
            score += 2
        }
        if (
            normalized.contains("meta name=\"csrf-token\"") ||
            normalized.contains("meta name='csrf-token'")
        ) {
            score += 1
        }
        return score
    }
}

private object FireBootstrapHtmlMetadataParser {
    fun currentUsername(html: String?): String? = metaContent("current-username", html)

    fun csrfToken(html: String?): String? = metaContent("csrf-token", html)

    private fun metaContent(name: String, html: String?): String? {
        if (html.isNullOrBlank()) {
            return null
        }

        val escapedName = Regex.escape(name)
        val patterns = listOf(
            """<meta\b[^>]*\bname\s*=\s*["']$escapedName["'][^>]*\bcontent\s*=\s*["']([^"']+)["'][^>]*>""",
            """<meta\b[^>]*\bcontent\s*=\s*["']([^"']+)["'][^>]*\bname\s*=\s*["']$escapedName["'][^>]*>""",
        )
        return patterns.firstNotNullOfOrNull { pattern ->
            Regex(pattern, RegexOption.IGNORE_CASE)
                .find(html)
                ?.groupValues
                ?.getOrNull(1)
                ?.trim()
                ?.takeIf { it.isNotEmpty() }
        }
    }
}
